proposals/NNNN-generalize-async-sequence.md

```diff
@@ -0,0 +1,311 @@
+* `param!` (the force-unwrapping operator); or
+* a reference to a _non-optional binding_ of `param`, i.e. a `let` constant initialized by a successful pattern-match which removes the optionality from `param`, such as `ref` in `if let ref = param`.
+
+When analyzing an argument expression in all cases above, certain non-instrumental differences in expression syntax and behavior must be ignored:
+
+* parentheses;
+* the effect-marking operators `try`, `try?`, `try!`, and `await`;
+* the type coercion operator as (in the cases where it doesn't perform a dynamic bridging conversion); and
+implicit type conversions such as promotion to `Optional` type.
+
+#### Calling `nextElement()` in async `for-in` loops
+
+Desugared async `for-in` loops will call `AsyncIteratorProtocol.nextElement()` instead of `next()` when the context has appropriate availability, and pass in an isolated argument value of type `(any Actor)?`. The argument value always represents the isolation of the caller so that the call does not cross an isolation boundary:
+
+* if the caller's context is `nonisolated`, the argument value is `nil`;
+* if the caller's context is isolated to some global actor `T`, the argument value is `T.shared`;
+* if the caller's context has an isolated parameter `param`, the argument value is `param`;
+* if the caller's context captures an isolated parameter `param`, the argument value is `param`.
+
+### Default implementations of `next()` and `nextElement()`
+
+Because existing `AsyncIteratorProtocol`-conforming types only implement `next()`, the standard library provides a default implementation of `nextElement()`:
+
+```swift
+extension AsyncIteratorProtocol {
+  /// Default implementation of `nextElement()` in terms of `next()`, which is
+  /// required to maintain backward compatibility with existing async iterators.
+  @available(SwiftStdlib 5.11, *)
+  @available(*, deprecated, message: "Provide an implementation of 'nextElement'")
+  public mutating func nextElement(_ actor: isolated (any Actor)?) async throws(Failure) -> Element? {
+    nonisolated(unsafe) var unsafeIterator = self
+    do {
+      let element = try await unsafeIterator.next()
+      self = unsafeIterator
+      return element
+    } catch {
+      throw error as! Failure
+    }
+  }
+}
+```
+
+Note that the default implementation of `nextElement` necessarily violates `Sendable` checking in order to pass `self` from a possibly-isolated context to a `nonisolated` one. Though this is generally unsafe, this is how calls to `next()` behave today, so existing conformances will maintain the behavior they already have. Implementing `nextElement()` directly will eliminate the unsafety.
+
+To enable conformances of `AsyncIteratorProtocol` to only implement `nextElement()`, a default implementation is also provided for `next()`:
+
+```swift
+extension AsyncIteratorProtocol {
+  @available(SwiftStdlib 5.11, *)
+  public mutating func next() async throws(Failure) -> Element? {
```

Comment:
The original `next()` function is not using typed throws, so for this function to be a valid default implementation of the `AsyncIteratorProtocol`'s `next()` requirement, we need to lose the `Failure` type.
```suggestion
  public mutating func next() async throws -> Element? {
```